src/main/kotlin/io/ionic/libs/ionfiletransferlib/IONFLTRController.kt

```diff
@@ -0,0 +1,343 @@
+            when {
+                options.url.isBlank() -> throw IONFLTRException.EmptyURL(options.url)
+                !inputsValidator.isURLValid(options.url) -> throw IONFLTRException.InvalidURL(options.url)
+                !inputsValidator.isPathValid(options.filePath) -> throw IONFLTRException.InvalidPath(options.filePath)
+            }
+
+            // Create parent directories if needed
+            val targetFile = File(options.filePath)
+            fileHelper.createParentDirectories(targetFile)
+
+            // Setup connection
+            val connection = setupConnection(options.url, options.httpOptions)
+            
+            try {
+                connection.connect()
+                
+                // Check response code
+                val responseCode = connection.responseCode
+                if (responseCode < 200 || responseCode > 299) {
+                    val errorBody = connection.errorStream?.bufferedReader()?.readText()
+                    throw IONFLTRException.HttpError(
+                        responseCode.toString(),
+                        errorBody,
+                        connection.headerFields.mapValues { it.value.firstOrNull() ?: "" }
+                    )
+                }
+                
+                // Get content length if available
+                val contentLength = connection.contentLength.toLong()
+                val lengthComputable = contentLength > 0
+                
+                // Download the file
+                BufferedInputStream(connection.inputStream).use { inputStream ->
+                    FileOutputStream(targetFile).use { fileOut ->
+                        BufferedOutputStream(fileOut).use { outputStream ->
+                            val buffer = ByteArray(BUFFER_SIZE)
+                            var bytesRead: Int
+                            var totalBytesRead: Long = 0
+                            
+                            while (inputStream.read(buffer).also { bytesRead = it } != -1) {
+                                outputStream.write(buffer, 0, bytesRead)
+                                totalBytesRead += bytesRead
+                                
+                                // Emit progress
+                                emit(
+                                    IONFLTRTransferResult.Ongoing(
+                                        IONFLTRProgressStatus(
+                                            bytes = totalBytesRead,
+                                            contentLength = contentLength,
+                                            lengthComputable = lengthComputable
+                                        )
+                                    )
+                                )
+                            }
+                            
+                            // Emit completion
+                            val headers = connection.headerFields.mapValues { it.value.firstOrNull() ?: "" }
```

Comment:
Thinking if we want to get just the first header value, information could be lost. I'd say either we change the `headers` property or convert the value list to string. Let me know if that makes sense, and if you have a preference.

```suggestion
                            val headers = connection.headerFields.mapValues { it.value.toString() }
```